8187229: Tree/TableCell: cancel event must return correct editing location

kleopatra opened the pull request:
the bug is an incorrect edit location (for Tree/Table: Tree/TablePosition) in edit cancel events - expected is the location at the time the cell edit was started, actual was the location of at the time the edit was cancelled. See the report for details.

The fix is analogue to those for ListCell/TreeCell, that is storing the edit location in startEdit and use that in cancelEdit.

Added tests that failed before and passed after and tests that (accidentally :) passed before and still pass after.

Related issues:

- also fixes [JDK-8269136](https://bugs.openjdk.java.net/browse/JDK-8269136) (Tree/TablePosition: must not throw NPE on instantiating with null table), in hind-sight it seemed too small to warrant its own PR. 
- does not fix the implementation of CellEditEvent (todo: file follow-up issue) which must not throw with null Tree/TablePosition.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8187229](https://bugs.openjdk.java.net/browse/JDK-8187229): Tree/TableCell: cancel event must return correct editing location
 * [JDK-8269136](https://bugs.openjdk.java.net/browse/JDK-8269136): Tree/TablePosition: must not throw NPE on instantiating with null table


### Reviewers
 * [Marius Hanl](https://openjdk.java.net/census#mhanl) (@Maran23 - Author)
 * [Ajit Ghaisas](https://openjdk.java.net/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/561/head:pull/561` \
`$ git checkout pull/561`

Update a local copy of the PR: \
`$ git checkout pull/561` \
`$ git pull https://git.openjdk.java.net/jfx pull/561/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 561`

View PR using the GUI difftool: \
`$ git pr show -t 561`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/561.diff">https://git.openjdk.java.net/jfx/pull/561.diff</a>

</details>

Maran23 commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TableCellTest.java:
```diff
@@ -402,6 +415,137 @@ public void testCellInUneditableColumnIsNotEditable() {
+        cell.updateTableColumn(editingColumn);
+    }
+
+
```
Very minor: 2 new lines

Maran23 commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TableCellTest.java:
```diff
@@ -402,6 +415,137 @@ public void testCellInUneditableColumnIsNotEditable() {
+     * takes care of wrapping everything into weakRefs.
+     */
+    @Test
+    public void testEditCancelMemoryLeakAfterRemoveEditingItem() {
```
I took a while to understand this test. 
Maybe one more sentence with information would be helpful here (and same test in TreeTableCell).

(My first thought was we are testing basically WeakReference, but then I figured out we want to test whether there are some references left somewhere which prevents the GC or not. So a small sentence like this would be nice in my opinion :))

kleopatra replied:
actually, I went the opposite direction: aligned all method comments to those in TreeCellTest (which boiled down to removing nearly all :)

Maran23 commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableCellTest.java:
```diff
@@ -708,6 +722,145 @@ public void testCellInUneditableColumnIsNotEditable() {
+    }
+
+    @Test
+    public void testEditCancelEventAfterModifyItems() {
```
Does it make sense to have the same javadoc here as the same test in TableCellTest has?

Maran23 commented on modules/javafx.controls/src/test/java/test/javafx/scene/control/TableCellTest.java:
```diff
@@ -402,6 +415,137 @@ public void testCellInUneditableColumnIsNotEditable() {
+     * Fails before fix of JDK-8187229.
+     */
+    @Test
+    public void testEditCancelEventAfterCancelOnTable() {
```
Minor: I think one more sentence about what to expect here would be good (so what is tested against here). Can be copied to all similar tests as well.

Maran23 reviewed (approved):
Fix looks good, I just saw the concept is the same as in your previous PR: https://github.com/openjdk/jfx/pull/539

Verified tests are failing before and passing now. (and as you mnetioned some pass before and pass now)

I left some inline comments, mostly minor and my opinion, so it's up to you. :)

aghaisas commented:
The fix is fine.
I have few suggestions-
1) Please add JDK-8269136 to this PR using "/issue add JDK-8269136" command
2) We do not have a precedent of adding comments to the tests indicating pass/fail before/after a bug. It is OK if some of them pass without this fix as well. Suggest to remove these type of comments from newly added tests.
3) Please file a follow on bug and update the todo: in PR description.

Something for the future :
I see that the fix of JDK-8269136 has its own separate code changes and tests. I see that you have combined it with this PR only due to the fact that the changes are smaller. I personally prefer NOT combining unrelated fixes. This ensures a better cohesion of the bug fixes. Added advantage is reverting a fix is easier in case of future regressions.

kleopatra commented:
/issue add JDK-8269136

kleopatra commented:
> 
> 
> The fix is fine.
> I have few suggestions-
> 
>     1. Please add JDK-8269136 to this PR using "/issue add JDK-8269136" command
> 

done

>     2. We do not have a precedent of adding comments to the tests indicating pass/fail before/after a bug. It is OK if some of them pass without this fix as well. Suggest to remove these type of comments from newly added tests.

argg .. you are right, sloppy me left more of them than I intended to (my locale branches are a bit crowded with comments ;) - aligned with TreeCellTest, which meant to remove most

> 
>     3. Please file a follow on bug and update the todo: in PR description.
> 

will do

> 
> Something for the future :
> I see that the fix of JDK-8269136 has its own separate code changes and tests. I see that you have combined it with this PR only due to the fact that the changes are smaller. I personally prefer NOT combining unrelated fixes. This ensures a better cohesion of the bug fixes. Added advantage is reverting a fix is easier in case of future regressions.

Fully agree with not co-fixing unrelated issues - just: I found it a bit hard to really separate them because they are not as unrelated as it looks: 

- fixing this introduces failing tests in CellTest without also fixing JDK-8269136 (because it seems to be allowed to switch a cell into editing state without it being attached to a control) 
- fixing JDK-8269136 without context (just the plain NPE) has no obvious relation to editing 

Hmm ... probably should update the description of this PR to explain that mix.

kleopatra commented:
> 
>     3. Please file a follow on bug and update the todo: in PR description.
> 

done - the follow-up is [JDK-8269871](https://bugs.openjdk.java.net/browse/JDK-8269871)

> I see that the fix of JDK-8269136 has its own separate code changes and tests. I see that you have combined it with this PR only due to the fact that the changes are smaller. 

removed that evaluation from the PR description, added the reasoning for inclusion (as outline in my previous comment) to [JDK-8269136](https://bugs.openjdk.java.net/browse/JDK-8269136)

Maran23 reviewed (approved):
looks good 👍

kleopatra commented:
/integrate
